rusttorney-server/src/server.rs

```diff
@@ -1,80 +1,188 @@
+    async fn handle_handshake(
         &mut self,
-        _hdid: String,
+        hdid: String,
     ) -> Result<(), anyhow::Error> {
-        self.socket.send(ServerCommand::Handshake(1111.to_string())).await?;
+        self.client.hdid = hdid.clone();
+        self.client_manager.lock().await.update_client(self.client.clone());
 
-        Ok(())
+        self.db.add_hdid(hdid, self.client.ipid).await?;
+
+        self.socket
+            .send(ServerCommand::ServerVersion(
+                self.client.id,
+                self.software.clone(),
+                self.version.clone(),
+            ))
+            .await?;
+
+        self.socket
+            .send(ServerCommand::PlayerCount(
+                self.player_count().await,
+                self.config.general.playerlimit,
+            ))
+            .await
+    }
+
+    async fn player_count(&self) -> u8 {
+        self.client_manager.lock().await.clients.iter().fold(0, |acc, c| {
+            if c.char_id != 1 {
+                acc + 1
+            } else {
+                acc
+            }
+        })
+    }
+
+    async fn handle_keepalive(&mut self) -> Result<(), anyhow::Error> {
+        self.ch_tx.send(()).await?;
+        self.socket.send(ServerCommand::KeepAlive).await
+    }
+
+    async fn start_handling(
+        mut self,
+        mut timeout_rx: Receiver<()>,
+    ) -> Result<(), anyhow::Error> {
+        // main client connection loop
+        loop {
+            // run concurrently timeout receiver and decoder, getting messages and handling them
+            select! {
+                _ = &mut timeout_rx => {
+                    return Err(anyhow::anyhow!("Client disconnected because of timeout!"));
+                }
+                res = self.socket.next() => {
+                    if let Some(parsed) = res {
+                        self.handle(parsed?).await?;
+                    } else {
+                        return Err(anyhow::anyhow!("Client disconnected!"));
+                    }
+                }
+            }
+        }
     }
 }
 
 impl<'a> AOServer<'a> {
-    pub fn new(
-        config: &'a Config<'a>,
-        db_pool: Pool<PostgresConnectionManager<NoTls>>,
-    ) -> anyhow::Result<Self> {
+    pub fn new(config: &'a Config<'a>, db: DbWrapper) -> anyhow::Result<Self> {
         let playerlimit = config.general.playerlimit;
         Ok(Self {
             config,
-            db_pool,
+            db: db.clone(),
             client_manager: Arc::new(Mutex::new(ClientManager::new(
                 playerlimit,
+                db,
             ))),
         })
     }
 
     async fn migrate(&mut self) -> anyhow::Result<()> {
-        let mut conn = self.db_pool.get().await?;
+        log::debug!("Getting pool connection for migration...");
+        let mut conn = self.db.get().await?;
         let stmt = conn.prepare("SELECT db_version FROM general_info").await;
 
         let current_version = match stmt {
```

Comment:
Why we are running migrations only if select errored? Shouldn't we also run them if current_version < latest_version?